src/actionspace/ObjectOrientedActionSpace.cpp

```diff
@@ -1,178 +1,157 @@
+//     2. Redistributions in binary form must reproduce the above copyright
+//        notice, this list of conditions and the following disclaimer in the
+//        documentation and/or other materials provided with the distribution.
+//     3. Neither the name of the copyright holder nor the names of its
+//        contributors may be used to endorse or promote products derived from
+//        this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
 ////////////////////////////////////////////////////////////////////////////////
 
+#include <iostream>
 #include "actionspace/ObjectOrientedActionSpace.hpp"
+#include "statespace/SE2.hpp"
+#include "utils/utils.hpp"
 
 namespace libcozmo {
 namespace actionspace {
 
+ObjectOrientedActionSpace::ObjectOrientedActionSpace(
+    const std::vector<double>& speeds,
+    const std::vector<double>& ratios,
+    const double& edge_offset,
+    const int& num_offset) : \
+    m_speeds(speeds),
+    m_ratios(ratios),
+    num_offset(num_offset),
+    m_center_offset(60),
+    m_edge_offset(edge_offset) {
+    std::vector<double> cube_offsets = num_offset == 1 ?
+        std::vector<double>{0} :
+        utils::linspace(-m_edge_offset, m_edge_offset, num_offset);
+    int action_id = 0;
+    for (const auto& heading_offset :
+        std::vector<double>{FRONT, LEFT, BACK, RIGHT}) {
+        double ratio = m_ratios[0];
+        if (heading_offset == LEFT || heading_offset == RIGHT) {
+            ratio = m_ratios[1];
+        }
+        for (const auto& cube_offset : cube_offsets) {
+            for (const auto& speed : speeds) {
+                ObjectOrientedActionSpace::GenericAction* action =
+                    static_cast<ObjectOrientedActionSpace::GenericAction*>(
+                        get_action(action_id));
+                m_actions.push_back(new GenericAction(
+                    speed,
+                    -cube_offset / m_edge_offset,
+                    ratio,
+                    heading_offset));
+                action_id++;
+            }
+        }
+    }
+}
+
 bool ObjectOrientedActionSpace::action_similarity(
     const int& action_id1,
```

Comment:
In the header file you should explain that action IDs correspond to generic actions and you can calculate the object oriented action for the ID if necessary. I would probably put this comment at the top of the class